src/backend/sgx/attestation.rs

```diff
@@ -1,17 +1,77 @@
 // SPDX-License-Identifier: Apache-2.0
 
+// Credit to: https://github.com/fortanix/rust-sgx/tree/master/aesm-client
+// for examples of AESM Requests.
+
+use crate::protobuf::aesm_proto::{
+    Request, Request_InitQuoteRequest, Response, Response_InitQuoteResponse,
+};
 use crate::syscall::{SGX_DUMMY_QUOTE, SGX_DUMMY_TI, SGX_QUOTE_SIZE, SGX_TI_SIZE};
 
-use std::io::Error;
+use std::io::{Error, Read, Write};
+use std::mem::size_of;
+use std::os::unix::net::UnixStream;
 use std::slice::from_raw_parts_mut;
 
+use protobuf::{Message, ProtobufResult};
+
+const AESM_SOCKET: &str = "/var/run/aesmd/aesm.socket";
+
 /// Fills the Target Info of the QE into the output buffer specified and
 /// returns the number of bytes written.
-// TODO: Replace synthetic TargetInfo with real TargetInfo.
-fn get_ti(out_buf: &mut [u8]) -> usize {
+fn get_ti(out_buf: &mut [u8]) -> Result<usize, std::io::Error> {
     assert_eq!(out_buf.len(), SGX_TI_SIZE, "Invalid size of output buffer");
-    out_buf.copy_from_slice(&SGX_DUMMY_TI);
-    SGX_TI_SIZE
+
+    // If unable to connect to the AESM daemon, return dummy value
+    let mut stream = match UnixStream::connect(AESM_SOCKET) {
+        Ok(s) => s,
+        Err(_) => {
+            out_buf.copy_from_slice(&SGX_DUMMY_TI);
+            return Ok(SGX_TI_SIZE);
+        }
+    };
+
+    // Set an Init Quote Request
+    let mut req = Request::new();
+    let mut msg = Request_InitQuoteRequest::new();
+    msg.set_timeout(1_000_000);
+    req.set_initQuoteReq(msg);
+
+    // Set up Writer
+    let mut buf_wrtr = vec![0u8; size_of::<u32>()];
+    match req.write_to_writer(&mut buf_wrtr) {
+        Ok(_) => {}
+        Err(e) => eprintln!("Error getting TargetInfo: {:#?}", e),
```

Comment:
Oh... good spot! I am curious also.

Reply:
I remember this being useful for debugging; for example, if the `req` wasn't formed correctly according to protobuf, it could be spotted early here instead of going through the other steps. But you're right that it should at least return early, or possibly be changed in other ways (open to more suggestions).